### ISSUE TITLE
Incorrect results returned from MATCH with `class` filter

### ISSUE THREAD

**Issue by obi1kenobi** [orientechnologies/orientdb#6845]
## OrientDB Version, operating system, or hardware.
- v2.0 SNAPSHOT[ ] - .18[ ] .17[ ] .16[ ] .15[ ] .14[ ] .13[ ] .12[ ] .11[ ] .10[ ] .9[ ] .8[ ] .7[ ] .6[ ] .5[ ] .4[ ] .3[ ] .2[ ] .1[ ] .0[ ]
- v2.1 SNAPSHOT[ ] - .16[ ] .15[ ] .14[ ] .13[ ] .12[ ] .11[ ] .10[ ] .9[ ] .8[ ] .7[ ] .6[ ] .5[ ] .4[ ] .3[ ] .2[ ] .1[ ] .0[ ]
- v2.2 SNAPSHOT[ ] - .8[x] .rc1[ ] .beta2[ ] .beta1[ ]

## Operating System
- [x] Linux
- [ ] MacOSX
- [ ] Windows
- [ ] Other Unix
- [ ] Other, name? 

On our production database, I am seeing incorrect results returned from a specific MATCH query involving a filtering step using the `class:` keyword after an edge traversal. Unfortunately, I was not able to construct a reproducible testcase that does not involve our proprietary data, and I am not able to share the data or the exact schema we are using.

Here is a schema and small dataset that capture some of the relevant relationships in our production setup:
```
CREATE CLASS Base EXTENDS V
CREATE PROPERTY Base.name String

CREATE CLASS AbstractFoo EXTENDS Base ABSTRACT
CREATE CLASS Foo EXTENDS AbstractFoo

CREATE CLASS AbstractMixin EXTENDS V ABSTRACT
CREATE CLASS GeneralBar EXTENDS Base
CREATE CLASS SpecificBar EXTENDS GeneralBar, AbstractMixin

CREATE CLASS Unrelated EXTENDS Base

CREATE CLASS AbstractFoo_Base EXTENDS E
CREATE PROPERTY AbstractFoo_Base.out LINK AbstractFoo
CREATE PROPERTY AbstractFoo_Base.in LINK Base

CREATE CLASS Unrelated_SpecificBar EXTENDS E
CREATE PROPERTY Unrelated_SpecificBar.out LINK Unrelated 
CREATE PROPERTY Unrelated_SpecificBar.in LINK SpecificBar

CREATE VERTEX Foo SET name = 'foo'
CREATE VERTEX SpecificBar SET name = 'bar'
CREATE VERTEX Unrelated SET name = 'unrelated'
CREATE EDGE AbstractFoo_Base FROM (SELECT FROM Foo) TO (SELECT FROM SpecificBar)
CREATE EDGE Unrelated_SpecificBar FROM (SELECT FROM Unrelated) TO (SELECT FROM SpecificBar)
```

## Expected behavior and actual behavior
Expected: queries that look like the following:
```
// query 1
MATCH {
    class: SpecificBar,
    as: origin,
    where: (name = 'bar')
}.in('AbstractFoo_Base') {
    class: Foo,
    as: endpoint
}
RETURN origin, endpoint
```
have the property that `origin instanceof SpecificBar` and `endpoint instanceof Foo` are both true, as the `class:` query keywords in the two match steps indicate.

Actual: we found a query on our production system for which one of the returned values was not an instance of the class with which it was supposed to be filtered in its match step. The query was similar to this:
```
// query 2
MATCH {
    class: SpecificBar,
    as: origin,
    where: (name = 'bar')
}.in('AbstractFoo_Base') {
    class: Unrelated,
    as: endpoint
}
RETURN origin, endpoint
```

Note that executing the above query on OrientDB 2.2.8, with the above schema and data, correctly returns no results. This is because the vertices of type `Unrelated` cannot and are not connected by an edge of type `AbstractFoo_Base`.

However, the analogous query to query 2 on our production OrientDB 2.2.8 returned one row -- the same row as query 1, for which `endpoint instanceof Unrelated` is not true as `endpoint` is of type `Foo`.
 
## Steps to reproduce the problem
As I mentioned, I was not able to reproduce the problem outside of our production system. However, I narrowed down the set of conditions that must be true for the bug to appear, and was hoping that someone with deeper knowledge of the MATCH execution system could use that to find the problem. Here is what I found:
- Replacing `class: Unrelated` with `class: E` has the same incorrect result.
- Replacing `class: Unrelated` with a class that does not exist generates an error (correctly).
- Replacing `class: Unrelated` with a class that has no entries that are connected via edges of any type to `SpecificBar`, correctly returns no results.
- Replacing `class: Unrelated` with a class `X` such that the schema specifies an edge that connects `X` to `SpecificBar`, but there are no entries of type `X` or its subtypes, correctly returns no results.
- Replacing `class: Unrelated` with any class `Y` that has entries connected by edges to a `SpecificBar` that is not the one selected by the first step in the `MATCH`, correctly returns no results.
- Replacing `class: Unrelated` with any class `Z` that has entries connected by edges to the exact `SpecificBar` that is selected by the first step in the `MATCH`, **generates the incorrect entry**, even though `Z` is not a superclass of `SpecificBar`.

One additional thing that may be complicating things -- we have kept the same database (upgraded from version to version) from approximately OrientDB 2.1.5, so the issue may also depend on that upgrade process. I was not able to export/import the data into a completely-new OrientDB database due to #6758 / #6759.

This is also a serious problem that is currently impeding my GraphQL-to-MATCH compilation work, as the compilation generates correct outputs that nevertheless produce incorrect results.

Thanks in advance for looking into it, and please let me know if I can help your investigation and debugging.

**Comment by obi1kenobi** [comment]
As a follow-up, I tried rewriting the query to replace the `class:` in the second traversal step with a `where: ($currentMatch instanceof Foo)`, but that seemed to **always evaluate to false**, regardless of whether the current match was actually of that class or not. I'm not sure if this is intended behavior or another bug.

**Comment by luigidellaquila** [comment]
Hi @obi1kenobi 

Thank you for reporting, I'll check it asap

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Hi @obi1kenobi 

I think I already fixed this problem, any chance to try it on 2.2.12?
I'm doing the same in the meantime

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Yep, I can confirm, it's already fixed in 2.2.12.
I'm closing it, please reopen if I'm missing something

Thanks

Luigi

**Comment by obi1kenobi** [comment]
Thanks, it appears it is fixed in 2.2.12, but I was not able to find it in the [changelog](https://github.com/orientechnologies/orientdb/milestone/105?closed=1).

**Comment by luigidellaquila** [comment]
mmm... right, probably we did not set the milestone on the related issue.
I'll see if I can find and fix it ;-)

Thanks

Luigi
